src/analysis/asymptotics/negligible.lean

```diff
@@ -0,0 +1,204 @@
+import data.polynomial.eval
+
+/-!
+# Negligible Functions
+
+This file defines a predicate `negligible f` on functions `f` from `ℕ` to a normed field `𝕜`.
+
+The main theorem is `negligible_polynomial_mul` that says the product of a polynomial
+  and a negligible function is still a negligible function
+-/
+
+namespace asymptotics
+
+/-- Definition of negligible functions over an arbitrary `normed_field`.
+  Note that the second function always has type `ℕ → ℝ`, which generally gives better lemmas. -/
+def negligible {𝕜 : Type*} [normed_ring 𝕜] (f : ℕ → 𝕜) :=
+∀ (c : ℤ), is_O f (λ n, (n : ℝ) ^ c) filter.at_top
+
+section normed_ring
+
+variables {R : Type*} [normed_ring R]
+variables {f g : ℕ → R}
+
+lemma negligible_of_is_O (hg : negligible g)
+  (h : is_O f g filter.at_top) : negligible f :=
+λ c, h.trans $ hg c
+
+lemma negligible_of_le (hg : negligible g)
+  (h : ∀ n, ∥f n∥ ≤ ∥g n∥) : negligible f :=
+negligible_of_is_O hg (is_O_of_le filter.at_top h)
+
+lemma negligible_of_eventually_le (hg : negligible g)
+  (h : ∀ᶠ n in filter.at_top, ∥f n∥ ≤ ∥g n∥) : negligible f :=
+negligible_of_is_O hg $ is_O_iff.2 ⟨1, by simpa only [one_mul] using h⟩
+
+/-- It suffices to check the negligiblity condition for only sufficiently small exponents `c`.
+  See `negligible_of_is_O_fpow_le` for a version with explicit bounds -/
+lemma negligible_of_eventually_is_O
+  (h : ∀ᶠ (c : ℤ) in filter.at_bot, is_O f (λ n, (n : ℝ) ^ c) filter.at_top) :
+  negligible f :=
+begin
+  obtain ⟨C, hC⟩ := filter.eventually_at_bot.mp h,
+  intro c,
+  by_cases hc : c ≤ C,
+  { exact hC c hc },
+  { refine (hC C le_rfl).trans (is_O.of_bound 1 (filter.eventually_at_top.2 ⟨1, (λ b hb, _)⟩)),
+    simp_rw [one_mul, normed_field.norm_fpow, real.norm_coe_nat],
+    have hb : 1 ≤ (b : ℝ) := le_trans (le_of_eq nat.cast_one.symm) (nat.cast_le.2 hb),
+    exact fpow_le_of_le hb (le_of_not_le hc) }
+end
+
+lemma negligible_of_is_O_fpow_le (C : ℤ)
+  (h : ∀ c ≤ C, is_O f (λ n, (n : ℝ) ^ c) filter.at_top) :
+  negligible f :=
+negligible_of_eventually_is_O (filter.eventually_at_bot.2 ⟨C, h⟩)
+
+lemma negligible_of_is_O_fpow_lt (C : ℤ)
+  (h : ∀ c < C, is_O f (λ n, (n : ℝ) ^ c) filter.at_top) :
+  negligible f :=
+negligible_of_is_O_fpow_le C.pred
+  (λ c hc, h c (lt_of_le_of_lt hc (int.pred_self_lt C)))
+
+/-- A negligible function must tend to zero in the base ring (not just in norm) -/
+lemma tendsto_zero_of_negligible (hf : negligible f) :
+  filter.tendsto f filter.at_top (nhds 0) :=
+begin
+  refine is_O.trans_tendsto (hf (-1)) _,
+  have : (has_inv.inv : ℝ → ℝ) ∘ (coe : ℕ → ℝ) = (λ (n : ℕ), (n : ℝ) ^ (-1 : ℤ)),
+  by simp only [gpow_one, fpow_neg],
+  exact this ▸ filter.tendsto.comp (tendsto_inv_at_top_zero) (coe_nat_tendsto_at_top ℝ),
+end
+
+/-- A negligible function eventually has norm less than any positive bound -/
+lemma norm_eventually_le_of_negligible
+  (hf : negligible f) (ε : ℝ) (hε : 0 < ε) :
+  ∀ᶠ (n : ℕ) in filter.at_top, ∥f n∥ ≤ ε :=
```

Comment:
This should immediately follow from `tendsto f at_top (𝓝 0)`. Something like (untested):
```lean
by simpa using hf.tendsto_zero.eventually (closed_ball_mem_nhds (0 : R) hε)
```